Support node kill

saxtot opened the pull request:
Necessary for UI launch and kill.

GLDuval committed:
Add shutdownCallback used on rosnode kill

saxtot committed:
:test_tube: Change node ros::spin logic

Go from using ros::spin to ros::spinOnce in a while loop with flags

saxtot committed:
:rotating_light: Add Ros Rate to prevent Ros spinOnce from taking over CPU

Since nothing else was in the while() loop, spinOnce was called infinitely

lvanasse commented:
Did it really help with the CPU consumption?

saxtot committed:
:wrench: Adjust rate value to reduce response time

Bump from 5Hz to 100Hz. Will confirm if it matches with usage.

saxtot commented:
Yes, before the change it used 100% of a core, now it hovers between unnoticeable and 0.7%.

lvanasse commented:
Awesome! Good job :)!

lvanasse commented:
Do you plan on adding the listen on the gpio256 for estop physical state, or you want to let Leo do it ?

saxtot commented:
I'll do it with him. I will start another branch to allow this PR to merge in the mean time.

lvanasse commented on src/estop.cpp:
```diff
@@ -37,6 +41,26 @@ void sigintShutdownNode(int sig)
+void shutdownCallback(XmlRpc::XmlRpcValue& params, XmlRpc::XmlRpcValue& result)
+{
+  int num_params = 0;
+  if (params.getType() == XmlRpc::XmlRpcValue::TypeArray)
```
I'm not a fan of implicit braces. But the ROS C++ Standard allow it, so I'll leave it up to you.

saxtot replied:
Addressed to match the rest of our codebase in commit [093f6dd](https://github.com/clubcapra/capra_estop/pull/3/commits/093f6dd730abc3697618692cc945e137590240b9)

lvanasse commented on src/estop.cpp:
```diff
@@ -37,6 +41,26 @@ void sigintShutdownNode(int sig)
+  int num_params = 0;
+  if (params.getType() == XmlRpc::XmlRpcValue::TypeArray)
+    num_params = params.size();
+  if (num_params > 1)
```
Instead of add num_params & etc, could we just use params.size() in the if condition ?

saxtot replied:
That does function properly on my pc, but XmlRpc::XmlRpcValue::TypeArray is from an enum containing 6 other possible types. As such, params could theoretically also be one of those type. I have no clue in what scenario that would happen though.

lvanasse replied:
Ah I see. That's fair. Thanks for clearing things up for me.

lvanasse reviewed (dismissed):
If you have good reason for the `num_params` than it looks good to me. But if not, I think it might be cleaner to remove the variables.

saxtot committed:
:art: Add small readability and quality changes

Remove hard coded value and changed format of imported section to match codebase

lvanasse commented:
Did you test it thoroughly ? I'm wondering if we should add `ros::ok()` in the while loop to also check if the roscore process is terminated that the estop node shut down too.

saxtot commented:
That is what I was currently investigating. Using rosnode kill does not terminate the roscore. However, this does not fall in our current use cases, since the issue being resolved in this PR is closing the node with rosnode kill with another node. As such, the roscore will already be running when the estop node is created and will continue to run after it is killed, because the ros_bridge and launch_handler are always running.
I wonder if it would be worth it to fix such a specific situation where you specifically manually rosnode kill the estop node  and would want the roscore to also be terminated, which never happened before since we CTRL+C 99.99% of the time.
According to my quick Google searches, the fastest solution would be the usage of rosnode kill --all to terminate the core.
[example](https://answers.ros.org/question/237862/rosnode-kill/)

lvanasse commented:
> That is what I was currently investigating. Using rosnode kill does not terminate the roscore. However, this does not fall in our current use cases, since the issue being resolved in this PR is closing the node with rosnode kill with another node. As such, the roscore will already be running when the estop node is created and will continue to run after it is killed, because the ros_bridge and launch_handler are always running. I wonder if it would be worth it to fix such a specific situation where you specifically manually rosnode kill the estop node and would want the roscore to also be terminated, which never happened before since we CTRL+C 99.99% of the time. According to my quick Google searches, the fastest solution would be the usage of rosnode kill --all to terminate the core. [example](https://answers.ros.org/question/237862/rosnode-kill/)

Awesome, sounds good to me. Thank you for answering it.

benmalenfant commented on src/estop.cpp:
```diff
@@ -56,11 +85,22 @@ int main(int argc, char **argv)
     initializeGPIO();
 
+    ros::XMLRPCManager::instance()->unbind("shutdown");
+    ros::XMLRPCManager::instance()->bind("shutdown", shutdownCallback);
```
Shouldn't we use the same code to handle sigint and ros initiated shutdowns?

saxtot replied:
It has been awhile, but I think that was the point of the PR: sigint and rosnode kill can't be handled in the same way and need their own code.

benmalenfant replied:
yeah but I was talking about having the shutdown behaviour integrated in a private function so that it can affect both shutdown with one modification, let's say in the future, we want to turn off the lights on shutdown, that you'd have to add it to both shutdown routines

lvanasse replied:
Wouldn't be possible to add whatever that we want to shut down to the shutdownCallback function ?

saxtot replied:
I think I get what you mean, but wouldn't that require a new function that would be called by both methods? Since sig and xml require different parameters in their shutdown functions.

benmalenfant replied:
> I think I get what you mean, but wouldn't that require a new function that would be called by both methods? Since sig and xml require different parameters in their shutdown functions.

yeah that's what I meant but it's really overkill to isolate 1 line of code

saxtot replied:
This is good to merge then?

benmalenfant replied:
yup

benmalenfant reviewed (commented):
Great! :D May be worth looking at merging the 2 shutdown types into one but it's just one line of code...

benmalenfant commented:
Also there is no error handling on the toggleEstopEnable/disable and that name is really confusing, you're not toggling it, you're enabling or disabling it, but this has nothing to do with this pr, just tought I'd take note
